cdap-security/src/main/java/io/cdap/cdap/security/guice/CoreSecurityRuntimeModule.java

```diff
@@ -51,10 +52,15 @@ public Module getDistributedModules() {
    * Returns {@code true} if a {@link ZKClient} binding is needed for the distributed module.
    */
   public static CoreSecurityModule getDistributedModule(CConfiguration cConf) {
-    // If the file based path is not set explicitly, use ZK.
+    // If security is not needed, we don't need a distributed security module.
+    // It is merely for satisfying the binding dependencies only.
+    if (!cConf.getBoolean(Constants.Security.ENABLED) && !cConf.getBoolean(Constants.Security.ENFORCE_INTERNAL_AUTH)) {
+      return new InMemoryCoreSecurityModule();
+    }
+
     // If ZK is set explicitly, always use ZK.
     // This is the backward compatible behavior.
-    if (ZKClientModule.requiresZKClient(cConf)) {
+    if (!Strings.isNullOrEmpty(cConf.get(Constants.Zookeeper.QUORUM))) {
```

Comment:
So the case in which QUORUM is not set is no longer defaulting to using ZK on localhost?

Reply:
Right. In short, checking for overridden doesn’t work for containers spawn by app fabric due to the serialization and recreation of cConf, which makes every config as overridden